src/main/java/javaslang/collection/HashTree.java

```diff
@@ -0,0 +1,404 @@
+/*     / \____  _    ______   _____ / \____   ____  _____
+ *    /  \__  \/ \  / \__  \ /  __//  \__  \ /    \/ __  \   Javaslang
+ *  _/  // _\  \  \/  / _\  \\_  \/  // _\  \  /\  \__/  /   Copyright 2014-2015 Daniel Dietrich
+ * /___/ \_____/\____/\_____/____/\___\_____/_/  \_/____/    Licensed under the Apache License, Version 2.0
+ */
+package javaslang.collection;
+
+import javaslang.Tuple;
+import javaslang.Tuple2;
+import javaslang.control.None;
+import javaslang.control.Option;
+
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.Iterator;
+
+/**
+ * <a href="https://en.wikipedia.org/wiki/Hash_array_mapped_trie">Hash array mapped trie (HAMT)</a>.
+ */
+abstract class HashTree<K, V> implements Iterable<Tuple2<K, V>> {
```

Comment:
Maybe we should name it `HashArrayMappedTrie` (as [specialization of `HashTree`](https://en.wikipedia.org/wiki/Hash_tree_(persistent_data_structure))). This is your original name. I will most probably also do this with `RedBlackTree`, which is a `BinarySearchTree`.